File: Paradigmas Lista/Python/Paradigmas_Lista.py
from abc import ABC, abstractmethod



class Carro:
    def __init__(self, marca, modelo, ano, velocidade, motor):
        self.motor = motor
        self.marca = marca
        self.modelo = modelo
        self.ano = ano
        self.velocidade = velocidade
    
    def exibir_detalhes(self):
        print(f"Marca: {self.marca}, Modelo: {self.modelo}, Ano: {self.ano}")
        
    def Acelerar(self):
        self.velocidade += 10
        
    def Freiar(self):
        self.velocidade -= 10
        if self.velocidade < 0:
            self.velocidade = 0
        
    def exibir_velocidade(self):
        print(f"Velocidade: {self.velocidade} km/h")

class Motor:
    def __init__(self, nome, potencia):
        self.potencia = potencia
        self.nome = nome
        
    def exibir_detalhes(self):
        return f"Motor de {self.potencia} HP, Tipo: {self.tipo}"
        

carro1 = Carro("Toyota", "Corolla", 2020, 40,"v8")
carro2 = Carro("Ford", "Mustang", 2018, 0,"v6")
carro3 = Carro("Chevrolet", "Camaro", 2021, 100,"v7")

print("Teste Carros")

carro1.Freiar()
carro2.Freiar()
carro3.Acelerar()
carro1.exibir_detalhes()
carro1.exibir_velocidade()
carro2.exibir_detalhes()
carro2.exibir_velocidade()
carro3.exibir_detalhes()
carro3.exibir_velocidade()

print("")
print("")

class ContaBancaria:
    def __init__(self, saldo, titular):
        self.__saldo = saldo
        self.titular = titular
    
    def get_saldo(self):
        return self.__saldo    

        
    def depositar(self, valor):
        if valor > 0:
            self.__saldo += valor
            print(f"Depósito de R${valor:.2f} realizado com sucesso.")
        else:
            print("Valor de depósito inválido.")
    
    
    def sacar(self, valor):
        if valor > self._saldo:
            raise SaldoInsuficienteException(
                f"Saldo insuficiente! Saldo atual: R${self._saldo:.2f}, valor solicitado: R${valor:.2f}"
            )
        elif valor <= 0:
            print("Valor de saque inválido.")
        else:
            self._saldo -= valor
            print(f"Saque de R${valor:.2f} realizado com sucesso.")


class SaldoInsuficienteException(Exception):
    def __init__(self, mensagem="Saldo insuficiente para realizar o saque."):
        super().__init__(mensagem)






class Animal:
    def __init__(self, nome):
        self.nome = nome
    
    def som(self):
        raise NotImplementedError

class Cachorro(Animal):
    def som(self):
        return f"{self.nome} faz: Au au!"

class Gato(Animal):
    def som(self):
        return f"{self.nome} faz: Miau!"
    
def Bicharada(animais):
    for animal in animais:
        print(animal.som())
        
animais = [Cachorro("Pandora"), Gato("igui"), Cachorro("Sky"), Gato("logan")]        
cachorro = Cachorro("Rex")
gato = Gato("Mingau")

print("Teste Animais")

print(cachorro.som())
print(gato.som())
Bicharada(animais)

print("")
print("")

class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []
        
    def adicionar_professor(self, professor):
        
        if professor not in self.professores:
            
            self.professores.append(professor)
            professor.adicionar_escola(self)
        
    def exibir_escola(self):
        
        print(f"Escola: {self.nome}")
        print("Professores:")
        
        for professor in self.professores:
            print(f"- {professor.nome}, Materia: {professor.materia}")

class Professor:
    def __init__(self, nome, materia):
        self.materia = materia
        self.nome = nome
        self.escolas = []
        
    def adicionar_escola(self, escola):
        
        if escola not in self.escolas:
            
            self.escolas.append(escola)
            escola.adicionar_professor(self)
            
    def exibir_professor(self):
        
        print(f"Professor: {self.nome}, Materia: {self.materia}")
        print("Escolas onde leciona:")
        
        for escola in self.escolas:
            print(f"- {escola.nome}")
            
escola = Escola("Unipe")
professor = Professor("Ricardo", "Paradigmas Lin. Programação")

escola.adicionar_professor(professor)

print("Teste Escola")

escola.exibir_escola()
professor.exibir_professor()

print("")
print("")

class Empresa:
    def __init__(self, nome):
        self.nome = nome
        self.empregados = []
        
    def adicionar_empregado(self, empregado):
        
        if empregado not in self.empregados:
            self.empregados.append(empregado)


    def exibir_empregados(self):
        
        print(f"Empresa: {self.nome}")
        print("Empregados:")
        
        for empregado in self.empregados:
            print(f"- {empregado.exibir_empregado()}")

class Empregado:
    def __init__(self, nome, cargo, salario):
        self.salario = salario
        self.nome = nome
        self.cargo = cargo
        
    def exibir_empregado(self):
        return f"Nome: {self.nome}, Cargo: {self.cargo}, Salário: R${self.salario:.2f}"
        
empresa = Empresa("DevTech")
empregado = Empregado("Luan alberto", "Desenvolvedor", 5500.00)

empresa.adicionar_empregado(empregado)

print("Teste Empresa")

empresa.exibir_empregados()

print("")
print("")


class Imprimivel(ABC):
    
    @abstractmethod
    def imprimir(self):
        pass
    
    
class Relatorio(Imprimivel):
    def __init__(self, titulo, conteudo):
        self.titulo = titulo
        self.conteudo = conteudo
    
    def imprimir(self):
        print(f"Relatório: {self.titulo}")
        print(f"Conteúdo: {self.conteudo}")
        
class Contrato(Imprimivel):
    def __init__(self, partes, objeto):
        self.partes = partes
        self.objeto = objeto
    
    def imprimir(self):
        print(f"Contrato entre: {self.partes}")
        print(f"Objeto do contrato: {self.objeto}")
      
      
def imprimir_documento(documento: Imprimivel):
    documento.imprimir()  
        
relatorio = Relatorio("Relatório Financeiro", "Conteúdo do relatório financeiro.")
contrato = Contrato("DevTech, Unipe", "Prestação de serviços de TI")

print("Teste Documento")

imprimir_documento(relatorio)
print()
imprimir_documento(contrato)

print()
print()


class Calculadora:
    
    def somar(self, *numeros):
        
        if len(numeros) == 2:
            return numeros[0] + numeros[1]
        elif len(numeros) == 3:
            return numeros[0] + numeros[1] + numeros[2]
        else:
            raise ValueError("O método somar aceita apenas 2 ou 3 números")
        
calc = Calculadora()

print("Teste Calculadora")

resultado_2_numeros = calc.somar(10, 23)
print(f"Soma de 2 números: {resultado_2_numeros}")

resultado_3_numeros = calc.somar(5, 12, 15)
print(f"Soma de 3 números: {resultado_3_numeros}")

print()
print()



class Funcionario(ABC):
    
    def __init__(self, nome):
        self.nome = nome
    
    @abstractmethod
    def calcularSalario(self):
        pass

class FuncionarioHorista(Funcionario):
    
    def __init__(self, nome, horas_trabalhadas, valor_hora):
        super().__init__(nome)
        
        self.horas_trabalhadas = horas_trabalhadas
        self.valor_hora = valor_hora
    
    def calcularSalario(self):
        return self.horas_trabalhadas * self.valor_hora
    
    
class FuncionarioAssalariado(Funcionario):
    
    def __init__(self, nome, salario_mensal):
        super().__init__(nome)
        
        self.salario_mensal = salario_mensal
    
    def calcularSalario(self):
        return self.salario_mensal
    
    
    
horista = FuncionarioHorista("Luan", 160, 50)
assalariado = FuncionarioAssalariado("Ricardo", 10000)

print("Teste Funcionarios")
print(f"Salário de {horista.nome}: R${horista.calcularSalario():.2f}")
print(f"Salário de {assalariado.nome}: R${assalariado.calcularSalario():.2f}")

print()
print()



class Produto:
    def __init__(self, nome, preco):
        
        self.nome = nome
        self.preco = preco

    def __add__(self, outro_produto):
        
        if isinstance(outro_produto, Produto):
            return self.preco + outro_produto.preco
    
    def __str__(self):
        return f"{self.nome}: R${self.preco:.2f}"
    
    
produto1 = Produto("Pizza Congelada", 22.50)
produto2 = Produto("Mouse Bluetooth", 150.75)

total = produto1 + produto2

print("Teste Produtos")
print(f"Soma dos preços de {produto1.nome} e {produto2.nome}: R${total:.2f}")



class Matematica:
    
    @staticmethod
    def fatorial(n):
        
        if n < 0:
            raise ValueError("Fatorial não é definido para números negativos.")
        elif n == 0 or n == 1:
            return 1
        else:
            fatorial = 1
            for i in range(2, n + 1):
                fatorial *= i
            return fatorial

numero = 5

resultado = Matematica.fatorial(numero)

print("Teste Matematica")
print(f"Fatorial de {numero} é: {resultado}")
print()
print()



class Configuracao:

    _instancia = None
    
    def __new__(cls):
        
        if cls._instancia is None:
            cls._instancia = super(Configuracao, cls).__new__(cls)
        return cls._instancia
    
    def __init__(self):
        if not hasattr(self, 'configuracoes'):

            self.configuracoes = {}
    
    def definir_configuracao(self, chave, valor):
        self.configuracoes[chave] = valor
    
    def obter_configuracao(self, chave):
        return self.configuracoes.get(chave, None)

config1 = Configuracao()
config2 = Configuracao()

print("Teste Configuração")

print(f"Config1 é config2? {config1 is config2}")

config1.definir_configuracao("API_URL", "https://api.exemplo.com")

print(f"Configuração API_URL em config2: {config2.obter_configuracao('API_URL')}")

File: Paradigmas Lista/Go/Paradigmas_Lista.go
package main

import (
	"fmt"
	"sync"
)

// 14. Singleton (Configuracao)
type Configuracao struct {
	dados map[string]string
}

var instance *Configuracao
var once sync.Once

func GetConfiguracao() *Configuracao {
	once.Do(func() {
		instance = &Configuracao{dados: make(map[string]string)}
	})
	return instance
}

func (c *Configuracao) DefinirConfiguracao(chave, valor string) {
	c.dados[chave] = valor
}

func (c *Configuracao) ObterConfiguracao(chave string) string {
	return c.dados[chave]
}

// 1. Carro com Motor (Composição)
type Motor struct {
	Potencia int
}

type Carro struct {
	Marca      string
	Modelo     string
	Ano        int
	Motor      Motor
	Velocidade int
}

func (c *Carro) ExibirDetalhes() {
	fmt.Printf("Carro: %s %s, Ano: %d, Potência do Motor: %d CV\n", c.Marca, c.Modelo, c.Ano, c.Motor.Potencia)
}

func (c *Carro) Acelerar(aumento int) {
	c.Velocidade += aumento
}

func (c *Carro) Frear(reducao int) {
	c.Velocidade -= reducao
}

func (c *Carro) ExibirVelocidade() {
	fmt.Printf("Velocidade atual do %s: %d km/h\n", c.Modelo, c.Velocidade)
}

// 3. Conta Bancaria
type ContaBancaria struct {
	titular string
	saldo   float64
}

func (cb *ContaBancaria) Depositar(valor float64) {
	cb.saldo += valor
}

func (cb *ContaBancaria) Sacar(valor float64) error {
	if valor > cb.saldo {
		return &SaldoInsuficienteError{Saldo: cb.saldo}
	}
	cb.saldo -= valor
	return nil
}

func (cb ContaBancaria) ExibirSaldo() float64 {
	return cb.saldo
}

func (cb ContaBancaria) GetTitular() string {
	return cb.titular
}

// 15. Exceção personalizada (SaldoInsuficiente)
type SaldoInsuficienteError struct {
	Saldo float64
}

func (e *SaldoInsuficienteError) Error() string {
	return fmt.Sprintf("Saldo insuficiente. Saldo disponível: R$%.2f", e.Saldo)
}

// 4. Polimorfismo e Herança
type Animal interface {
	Som() string
}

type Cachorro struct {
	Nome string
}

func (c Cachorro) Som() string {
	return c.Nome + " faz: Au au!"
}

type Gato struct {
	Nome string
}

func (g Gato) Som() string {
	return g.Nome + " faz: Miau!"
}

func EmitirSons(animais []Animal) {
	for _, animal := range animais {
		fmt.Println(animal.Som())
	}
}

// 7. Associação (Escola e Professor)
type Professor struct {
	Nome    string
	Escolas []*Escola
}

type Escola struct {
	Nome        string
	Professores []*Professor
}

func (e *Escola) AdicionarProfessor(p *Professor) {
	e.Professores = append(e.Professores, p)
	p.Escolas = append(p.Escolas, e)
}

func (e *Escola) ExibirProfessores() {
	fmt.Printf("Escola %s tem os seguintes professores:\n", e.Nome)
	for _, professor := range e.Professores {
		fmt.Println(professor.Nome)
	}
}

func (p *Professor) ExibirEscolas() {
	fmt.Printf("O professor %s ensina nas seguintes escolas:\n", p.Nome)
	for _, escola := range p.Escolas {
		fmt.Println(escola.Nome)
	}
}

// 8. Agregação (Empresa e Empregados)
type Empregado struct {
	Nome    string
	Cargo   string
	Salario float64
}

type Empresa struct {
	Nome       string
	Empregados []Empregado
}

func (e *Empresa) AdicionarEmpregado(emp Empregado) {
	e.Empregados = append(e.Empregados, emp)
}

func (e *Empresa) ExibirEmpregados() {
	fmt.Printf("Empregados da empresa %s:\n", e.Nome)
	for _, empregado := range e.Empregados {
		fmt.Printf("Nome: %s, Cargo: %s, Salário: R$%.2f\n", empregado.Nome, empregado.Cargo, empregado.Salario)
	}
}

// 9. Interface Imprimível
type Imprimivel interface {
	Imprimir()
}

type Relatorio struct {
	Titulo   string
	Conteudo string
}

func (r Relatorio) Imprimir() {
	fmt.Printf("Relatório: %s\nConteúdo: %s\n", r.Titulo, r.Conteudo)
}

type Contrato struct {
	Termos string
}

func (c Contrato) Imprimir() {
	fmt.Printf("Contrato: %s\n", c.Termos)
}

func imprimirDocumento(doc Imprimivel) {
	doc.Imprimir()
}

// Main para execução
func main() {
	// Singleton
	config1 := GetConfiguracao()
	config2 := GetConfiguracao()
	fmt.Println("Config1 é config2?", config1 == config2)
	config1.DefinirConfiguracao("API_URL", "https://api.exemplo.com")
	fmt.Println("Configuração API_URL em config2:", config2.ObterConfiguracao("API_URL"))

	// Carro e Motor
	fmt.Println("\n Teste do Carro ")
	motor := Motor{Potencia: 120}
	carro := Carro{Marca: "Toyota", Modelo: "Corolla", Ano: 2020, Motor: motor}
	carro.ExibirDetalhes()
	carro.Acelerar(50)
	carro.ExibirVelocidade()
	carro.Frear(20)
	carro.ExibirVelocidade()

	// Conta Bancaria
	fmt.Println("\n Teste da Conta Bancária ")
	conta := ContaBancaria{titular: "João", saldo: 1000}
	fmt.Println("Titular:", conta.GetTitular())
	fmt.Println("Saldo inicial: R$", conta.ExibirSaldo())
	conta.Depositar(500)
	fmt.Println("Saldo após depósito: R$", conta.ExibirSaldo())
	err := conta.Sacar(300)
	if err != nil {
		fmt.Println("Erro:", err)
	} else {
		fmt.Println("Saldo após saque: R$", conta.ExibirSaldo())
	}
	err = conta.Sacar(1500)
	if err != nil {
		fmt.Println("Erro:", err)
	} else {
		fmt.Println("Saldo final: R$", conta.ExibirSaldo())
	}

	// Polimorfismo com animais
	fmt.Println("\n Sons dos Animais ")
	cachorro1 := Cachorro{Nome: "Pandora"}
	gato1 := Gato{Nome: "Mingau"}
	animais := []Animal{cachorro1, gato1}
	EmitirSons(animais)

	// Associação Escola e Professor
	fmt.Println("\n Associação Escola e Professor ")
	professor1 := Professor{Nome: "Ricardo"}
	professor2 := Professor{Nome: "Leonardo"}
	escola1 := Escola{Nome: "Unipe"}
	escola2 := Escola{Nome: "Uniesp"}
	escola1.AdicionarProfessor(&professor1)
	escola1.AdicionarProfessor(&professor2)
	escola2.AdicionarProfessor(&professor2)
	escola1.ExibirProfessores()
	escola2.ExibirProfessores()
	professor1.ExibirEscolas()
	professor2.ExibirEscolas()

	// Empresa e Empregados
	fmt.Println("\n Teste da Empresa ")
	empresa := Empresa{Nome: "Tech Solutions"}
	emp1 := Empregado{Nome: "Lucas", Cargo: "Desenvolvedor", Salario: 5000.0}
	emp2 := Empregado{Nome: "Mariana", Cargo: "Gerente de Projetos", Salario: 7000.0}
	empresa.AdicionarEmpregado(emp1)
	empresa.AdicionarEmpregado(emp2)
	empresa.ExibirEmpregados()

	// Interface Imprimível
	fmt.Println("\n Imprimindo Relatório e Contrato ")
	relatorio := Relatorio{Titulo: "Relatório Mensal", Conteudo: "Conteúdo do relatório..."}
	contrato := Contrato{Termos: "Cláusulas do contrato..."}
	imprimirDocumento(relatorio)
	imprimirDocumento(contrato)
}
